code/modules/chemistry/chemicompiler_core.dm

```diff
@@ -885,4 +655,37 @@
 		beepCode(3, 1) // No reservoir loaded in specified position
 		return 0
 	var/obj/item/reagent_containers/holder = reservoirs[rid]
+	tgui_process.update_uis(src.holder)
 	return holder.reagents.total_volume
+
+/datum/chemicompiler_executor/proc/get_ui_data()
+	. = core.get_ui_data()
+	var/list/ui_data_reservoirs = list()
+	for (var/i = 1, i <= length(src.reservoirs), i++)
+		var/obj/item/reagent_containers/R = src.reservoirs[i]
+		if (istype(R))
+			ui_data_reservoirs.Add("[R.reagents.total_volume]/[R.reagents.maximum_volume]")
+		else
+			ui_data_reservoirs.Add(null)
+	.["reservoirs"] = ui_data_reservoirs
+
+/datum/chemicompiler_executor/proc/execute_ui_act(action, list/params)
+	switch(action)
+		if("reservoir")
+			reservoirClick(params["index"] + 1)
+			return TRUE
+		if("save")
+			core.buttons[params["index"] + 1] = core.output
+			core.cbf[params["index"] + 1] = core.parseCBF(core.output, params["index"] + 1)
+			return TRUE
+		if("load")
+			core.output = core.buttons[params["index"] + 1]
+			core.loadTimestamp = world.timeofday
```

Comment:
I never remember how DM timestamps work, not saying this is wrong but commenting to sanity check if this is the correct one to use.

Reply:
TIME.